File: src/main/java/com/github/sblundy/elvish/psi/ElvishLambdaBase.java
package com.github.sblundy.elvish.psi;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class ElvishLambdaBase extends ASTWrapperPsiElement implements ElvishVariableScope {
    public ElvishLambdaBase(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public @NotNull Collection<ElvishVariableDeclaration> findVariables(@NotNull String name, @NotNull List<String> ns) {
        List<ElvishVariableDeclaration> found = new ArrayList<>();
        if (getLambdaArguments() != null && ns.isEmpty()) {
            for (ElvishVariableDeclaration declaration : getLambdaArguments().getParameterList()) {
                if (declaration.textMatches(name)) {
                    found.add(declaration);
                }
            }
        }

        if (!found.isEmpty()) {
            return found;
        }

        return ElvishPsiUtils.INSTANCE.findVariableInParentScope(name, ns, this);
    }

    @Nullable
    abstract public ElvishLambdaArguments getLambdaArguments();
}

File: src/main/kotlin/com/github/sblundy/elvish/psi/ElvishLambdaBase.kt
package com.github.sblundy.elvish.psi

import com.github.sblundy.elvish.psi.ElvishPsiUtils.findVariableInParentScope
import com.github.sblundy.elvish.psi.ElvishPsiUtils.processVariablesInParentScope
import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode

abstract class ElvishLambdaBase(node: ASTNode) : ASTWrapperPsiElement(node), ElvishVariableScope {
    override fun findVariables(ref: ReferenceWithNamespacePsiElement): Collection<ElvishVariableDeclaration> {
        val found = lambdaArguments?.let { lambdaArguments ->
            if (!ref.hasNamespace) {
                lambdaArguments.parameterList.filter { it.nameMatches(ref) }
            } else {
                listOf()
            }
        } ?: listOf()

        return if (found.isNotEmpty()) { found } else findVariableInParentScope(ref, this)
    }

    override fun processVariables(processor: ElvishVariableScope.VariableProcessor) {
        lambdaArguments?.parameterList?.let { parameterList ->
            val c = parameterList.any {
                processor.process(it)
            }
            if (c) {
                processVariablesInParentScope(processor, this)
            }
        }
    }

    abstract val lambdaArguments: ElvishLambdaArguments?
}